### DISCUSSION TITLE
Error while generating new case compset - F20TR_MMF1

### DISCUSSION THREAD

**Discussion by gourikadam** [E3SM-Project/E3SM#6699] [E3SM model help]
Dear team,

I am getting below mentioned error in ./case.build and the message says

```
/scratch/um_cdac2/met_models/CDACP_Int/E3SM/components/mpas-framework/src/external/ezxml/ezxml.c(647): error: identifier "MADV_SEQUENTIAL" is undefined
          madvise(m, l, MADV_SEQUENTIAL); // optimize for sequential access
                        ^

/scratch/um_cdac2/met_models/CDACP_Int/E3SM/components/mpas-framework/src/external/ezxml/ezxml.c(649): error: identifier "MADV_NORMAL" is undefined
          madvise(m, root->len = l, MADV_NORMAL); // put it back to normal
                                    ^

/scratch/um_cdac2/met_models/CDACP_Int/E3SM/components/mpas-framework/src/external/ezxml/ezxml.c(647): warning #266: function "madvise" declared implicitly
          madvise(m, l, MADV_SEQUENTIAL); // optimize for sequential access
          ^

/scratch/um_cdac2/met_models/CDACP_Int/E3SM/components/mpas-framework/src/external/ezxml/ezxml.c(647): error: identifier "MADV_SEQUENTIAL" is undefined
          madvise(m, l, MADV_SEQUENTIAL); // optimize for sequential access
                        ^

/scratch/um_cdac2/met_models/CDACP_Int/E3SM/components/mpas-framework/src/external/ezxml/ezxml.c(649): error: identifier "MADV_NORMAL" is undefined
          madvise(m, root->len = l, MADV_NORMAL); // put it back to normal
                                    ^

/scratch/um_cdac2/met_models/CDACP_Int/E3SM/components/mpas-framework/src/external/ezxml/ezxml.c(935): warning #266: function "strdup" declared implicitly
              xml->attr[1] = strdup(""); // empty list of malloced names/vals
```

Please help me in resolving this.

Thank you
Regards,
Gouri

**Comment by whannah1** [comment]
@gourikadam what machine/compiler are you trying to use?

**Reply by gourikadam** [reply]
Hi. I am using unsupported machine with intel compiler. my cmake file looks like below 
"if (COMP_NAME STREQUAL gptl)
  string(APPEND CMAKE_C_FLAGS " -DHAVE_SLASHPROC")
endif()
string(APPEND CMAKE_C_FLAGS " -std=c99")
if (compile_threaded)
  string(APPEND CMAKE_C_FLAGS   " -qopenmp")
  string(APPEND CMAKE_CXX_FLAGS " -qopenmp")
  string(APPEND CMAKE_Fortran_FLAGS   " -qopenmp")
  string(APPEND CMAKE_EXE_LINKER_FLAGS  " -qopenmp")
endif()
string(APPEND CMAKE_C_FLAGS_RELEASE   " -O2 -debug minimal")
string(APPEND CMAKE_CXX_FLAGS_RELEASE " -O2")
string(APPEND CMAKE_Fortran_FLAGS_RELEASE   " -O2 -debug minimal")
string(APPEND CMAKE_C_FLAGS_DEBUG   " -O0 -g")
string(APPEND CMAKE_CXX_FLAGS_DEBUG " -O0 -g")
string(APPEND CMAKE_Fortran_FLAGS_DEBUG   " -O0 -g -check uninit -check bounds -check pointers -fpe0 -check noarg_temp_created -init=snan,arrays")
#string(APPEND CMAKE_Fortran_FLAGS_DEBUG   " -L/scratch/um_cdac2/met_models/CDACP_Int/all-dep/lib -lpnetcdf -O0 -g -check uninit -check bounds -check pointers -fpe0 -check noarg_temp_created -init=snan,arrays")
string(APPEND CMAKE_CXX_FLAGS " -std=c++17 -xHost -fp-model fast")
if (COMP_NAME STREQUAL cice)
  string(APPEND CMAKE_Fortran_FLAGS_DEBUG " -init=nosnan,arrays")
endif()
if (COMP_NAME MATCHES "^pio")
  string(APPEND SPIO_CMAKE_OPTS " -DPIO_ENABLE_TOOLS:BOOL=ON")
endif()
string(APPEND CPPDEFS " -DFORTRANUNDERSCORE -DNO_R16 -DCPRINTEL")
string(APPEND CMAKE_Fortran_FLAGS " -convert big_endian -assume byterecl -ftz -traceback -assume realloc_lhs")
string(APPEND CMAKE_Fortran_FORMAT_FIXED_FLAG " -fixed -132")
string(APPEND CMAKE_Fortran_FORMAT_FREE_FLAG " -free")
string(APPEND LDFLAGS " ")
set(HAS_F2008_CONTIGUOUS "TRUE")
set(MPICC "mpiicc")
set(MPICXX "mpiicpc")
set(MPIFC "mpiifort")
set(SCC "icc")
set(SCXX "icpc")
set(SFC "ifort")
#set(LDFLAGS "-L/scratch/um_cdac2/met_models/CDACP_Int/all-dep/lib -lnetcdf -lpnetcdf")

# Intel has to link fortran mains with fortran
set(E3SM_LINK_WITH_FORTRAN "TRUE")
"

**Comment by whannah1** [comment]
Also, a few other sanity check things - are you using a recent version of the E3SM repo?
Did you make sure the submodules are synced and up-to-date?

**Reply by gourikadam** [reply]
Yes, I am using the recent version of E3SM downloaded from Github. As and when error occurs, installing respective submodules. Hopefully everything is updated.

**Comment by rljacob** [comment]
still looks like some submodule isn't checked out completely.  At the top of the code, do "git submodule update --init --recursive".  If you get an error, repeat until you don't.

**Reply by gourikadam** [reply]
I tried using  "git submodule update --init --recursive" but it is not working as access issue from the server and  i downloaded the original code using git clone. 
Cloning into 'components/eam/src/physics/cosp2/external'..                                                                                             .
Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
Clone of 'git@github.com:CFMIP/COSPv2.0.git' into submodul                                                                                             e path 'components/eam/src/physics/cosp2/external' failed

I manuaaly tried to clone few submodules  in the path "E3SM/cime/CIME"like cprnc. But still error is there. I have following doubts
1. wether cmake file which I posted in reply above is correct or any setting is missing
2. the Compset "F20TR_MMF1" which I am using for creating new EAM is correct or not?

**Reply by rljacob** [reply]
You either need to set up ssh keys with github or do "git config --global url."https://github.com/".insteadOf "git@github.com:"

**Reply by gourikadam** [reply]
Hi,

Thank you foryour support. Able to resolve most of the errors related to submodules. Now getting errro in e3sm.bld.log as below 
" elm/src/main/elm_interface_pflotranMod.F90.o.provides.build
/tmp/iforthpJeIj.i90: catastrophic error: **Internal compiler error: segmentation violation signal raised** Please report this error along with the circumstances in which it occurred in a Software Problem Report.  Note: File and line given may not be explicit cause of this error.
compilation aborted for /scratch/um_cdac2/met_models/CDACP_Int/E3SM/components/elm/src/external_models/fates/main/FatesHistoryInterfaceMod.F90 (code 1)
Target CMakeFiles/lnd.dir/__/__/elm/src/external_models/fates/main/FatesHistoryInterfaceMod.F90.o built in 2.777128 seconds
Target CMakeFiles/lnd.dir/__/__/elm/src/external_models/fates/fire/SFMainMod.F90.o built in 2.484418 seconds
make[2]: *** [cmake/lnd/CMakeFiles/lnd.dir/__/__/elm/src/external_models/fates/main/FatesHistoryInterfaceMod.F90.o] Error 1
"

**Comment by gourikadam** [comment]
![Screenshot 2024-11-04 120105_1](https://github.com/user-attachments/assets/df7adfaa-e8f6-46ef-b9f1-ced862780b6b)

**Reply by gourikadam** [reply]
After installing submosules using "git config --global url."[https://github.com/".insteadOf](https://github.com/%22.insteadOf) ", many submodules got installed. Still getting below error in e3sm bldlog
"catastrophic error: **Internal compiler error: segmentation violation signal raised** Please report this error along with the circumstances in which it occurred in a Software Problem Report.  Note: File and line given may not be explicit cause of this error."
I am using cmake3.27.9 and gcc 7.5, but still e3sm.bldlog shows 
"/home/apps/spack/opt/spack/linux-centos7-skylake_avx512/gcc-12.2.0/cmake-3.27.9-kn2wzltzm5rhboded5xbhqbxv22ynp2f/bin/cmake -E touch eamxx/src/mct_coupling/CMakeFiles/theta-l_kokkos_4_72_10.dir/__/__/__/homme/src/theta-l/share/element_ops.F90.o.provides.build
"
which means it is pointing to gcc 12.2.0. When gcc is upgraded to 12.2.0 its is giving different errors related to c++17. 

Please help regarding this.

Regards,
Gouri

**Reply by whannah1** [reply]
This seems to be a machine configuration issue. Resolving these things can sometimes take week of trial and error, and we can't provide specific guidance on how to resolve machine specific problems. Do you have `-std=c++17` in your flags? My only other suggestion is to keep trying different combinations of compiler and libraries and make sure the libraries like netcdf are built with the same compiler.

**Comment by mt5555** [comment]
Note that the MMF case you are using (superparameterized atmospehre) is significantly more complicated to build and run (and not an officially supported configuration of E3SM).  You might want to consider first getting the standard F20TR compset working (or even simpler, F2010)

**Reply by gourikadam** [reply]
Thank you very much.  These discussions and your remarks are really helpful. I have tried other compset, RMOSQIAN  and it worked. I will try F20TR and F2010 as suggested.
